lib/AST/ASTStructuralEquivalence.cpp

```diff
@@ -1674,6 +1785,17 @@ bool StructuralEquivalenceContext::CheckCommonEquivalence(Decl *D1, Decl *D2) {
 
   // FIXME: Move check for identifier names into this function.
 
+  if (auto *ND1 = dyn_cast<NamedDecl>(D1)) {
+    auto *ND2 = dyn_cast<NamedDecl>(D2);
+    if (!ND2)
+      return false;
```

Comment:
What if here we would just get the two DeclContexts from D1 and D2 and we would compare them by either NamedDecl::getQualifiedNameAsString or by index::generateUSRForDecl?
The USR seems more precise and both seems to be slow.
About the slowness we could add yet another local cache to the StructuralEquivalenceContext class.

How is it different than what we had previously in #624 ?
This time we would compare only the containing scopes by string and not the Decls themselves.

Reply:
What I am saying here is just an alternative solution to consider, not necessarily better ...

Reply:
I want to compare only scope, not the decl name itself (that is done already). Probably this was the reason for some test failures  when the qualified name was used.